### PR TITLE
[FLINK-21942][coordination] Remove job from LeaderIdService when disconnecting JobManager with globally terminal state

### PR DESCRIPTION
**Pull request by wangyang0918**
## What is the purpose of the change

We have the leader ConfigMap watch leak currently. When the job reaches to terminal state, the jobmanager leader retrieval service in `ResourceManager` is not stopped correctly. We start the "job leader id monitoring" in `ResourceManager#registerJobManager`, but we do not stop it when we `disconnectJobManager`.

This PR tries to remove the job with terminal state from `jobLeaderIdService` when `disconnectJobManager`.

## Brief change log

* Remove job from LeaderIdService when disconnecting JobManager with globally terminal stat

## Verifying this change

* Add two unit tests
* Verify manually in a native K8s session with K8s HA enabled. The following logs should be found when job is cancelled.
```
2021-03-29 07:00:15,048 INFO  org.apache.flink.kubernetes.highavailability.KubernetesLeaderRetrievalDriver [] - Stopping KubernetesLeaderRetrievalDriver{configMapName='k8s-session-ha-1-ca719ad578c9bf940ce478a98910ac0c-jobmanager-leader'}.
```

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wangyang0918** [comment]
cc @tillrohrmann could you please have a look this PR?

**Comment by wangyang0918** [comment]
Rebase latest master.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @wangyang0918. It goes in a good direction and I like the tests. I had a few comments for the thread safety of the tests and a suggestion for the `disconnectJobManager` code.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java [hunk 1]
```diff
@@ -509,6 +509,11 @@ public void disconnectTaskManager(final ResourceID resourceId, final Exception c
     @Override
     public void disconnectJobManager(
             final JobID jobId, JobStatus jobStatus, final Exception cause) {
+
+        if (jobStatus.isGloballyTerminalState()) {
```
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:513]: Yes. The only difference is we also check the existence in `jobManagerRegistrations` when `jobStatus` is globally terminal. And it is reasonable doing that.

**Comment by wangyang0918** [comment]
@tillrohrmann Thanks for your review. Comments addressed and force pushed.

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating this PR @wangyang0918. LGTM. The failing Mesos e2e test seems to be unrelated. Merging this PR.

**Pull request closed without merging**